src/pkg/clouds/aws/ecs/event.go

```diff
@@ -18,23 +38,297 @@ type ECSDeploymentStateChange struct {
+	eventCommonFields
+	Detail ECSTaskStateChange
+}
+
+type KanikoTaskStateChangeEvent TaskStateChangeEvent
+
+type ServiceActionEvent struct {
+	eventCommonFields
+	Detail ECSServiceAction
+}
+
+type DeploymentStateChangeEvent struct {
+	eventCommonFields
+	Detail ECSDeploymentStateChange
 }
 
-func FindContainerOverrides(detail *ECSTaskStateChange, containerName string) *ecstaskstatechange.OverridesItem {
-	for _, co := range detail.Overrides.ContainerOverrides {
-		if strings.HasPrefix(co.Name, containerName) {
-			return &co
+func ParseECSEvent(b []byte) (Event, error) {
+	var e struct {
+		eventCommonFields
+		Detail json.RawMessage `json:"detail"`
+	}
+	if err := json.Unmarshal(b, &e); err != nil {
+		return nil, err
+	}
+	switch e.DetailType {
+	case "ECS Task State Change":
+		var detail ECSTaskStateChange
+		if err := json.Unmarshal(e.Detail, &detail); err != nil {
+			return nil, err
+		}
+
+		var evt Event
+		for _, o := range detail.Overrides.ContainerOverrides {
+			if o.Name == "kaniko" {
+				evt = &KanikoTaskStateChangeEvent{e.eventCommonFields, detail}
+			}
+		}
+		if evt == nil {
+			e := &TaskStateChangeEvent{e.eventCommonFields, detail}
+			if len(e.Detail.Overrides.ContainerOverrides) == 0 {
+				return nil, fmt.Errorf("no container overrides for task state change event: %v", e.Detail.TaskArn)
+			}
+			evt = e
+		}
+
+		etag := evt.Etag()
+		deploymentId := detail.StartedBy
+		if etag != "" && deploymentId != "" {
+			DeploymentEtags.Set(deploymentId, etag)
+		}
+
+		return evt, nil
+	case "ECS Service Action":
+		var detail ECSServiceAction
+		if err := json.Unmarshal(e.Detail, &detail); err != nil {
+			return nil, err
+		}
+		return &ServiceActionEvent{e.eventCommonFields, detail}, nil
+	case "ECS Deployment State Change":
+		var detail ECSDeploymentStateChange
+		if err := json.Unmarshal(e.Detail, &detail); err != nil {
+			return nil, err
+		}
+		return &DeploymentStateChangeEvent{e.eventCommonFields, detail}, nil
+	default:
+		return nil, fmt.Errorf("unsupported ECS event type: %s", e.DetailType)
+	}
+}
+
+func (e *TaskStateChangeEvent) Service() string {
+	var service string
+	override := e.Detail.Overrides.ContainerOverrides[0]
+	i := strings.LastIndex(override.Name, "_")
+	if i > 0 {
+		service = override.Name[:i]
+	}
+	return service
+}
+func (e *TaskStateChangeEvent) Etag() string {
+	var etag string
+	override := e.Detail.Overrides.ContainerOverrides[0]
+	i := strings.LastIndex(override.Name, "_")
+	if i > 0 {
+		etag = override.Name[i+1:]
+	}
+	return etag
+}
+func (e *TaskStateChangeEvent) Host() string {
+	return path.Base(e.Detail.TaskArn)
+}
+
+func (e *TaskStateChangeEvent) Status() string {
+	status := "TASK_" + e.Detail.LastStatus
+	if e.Detail.StoppedReason != "" {
+		status += " " + e.Detail.StoppedReason
+	}
+	return status
+}
+
+func (e *TaskStateChangeEvent) State() defangv1.ServiceState {
+	state := defangv1.ServiceState_NOT_SPECIFIED
+	if e.Detail.LastStatus == "DEACTIVATING" {
```

Comment:
move out to make it more explicit

Reply:
Offline discussion:
- ServiceState should not have task oriented states like "TASK_FAIL"
- Which means we need to parse the status string at the cli subscribe level, which is less desirable
As a result: keep the current implementation.